### DISCUSSION TITLE
Version Control/Dolt app compatibility

### DISCUSSION THREAD

**Discussion by tim-fiola** [nautobot/pynautobot#26] [Ideas]
With default behavior, pynautobot does not fully work with querying Nautobot running the Version Control app because pynautobot, by default, will only query against the main/default/production branch.
 
With a simple mod, pynautobot can be made to query a specific, non-main/non-default Version Control branch.  The mod in question involves adding a `dolt-branch` key, value to the pynautobot api header, as such: `nautobot_api.http_session.headers['dolt-branch'] = "<non-default-branch-name>"`

Take this example below, where pynautobot is configured to query the 'atl-leaf-augment' branch for a device that does not exist in the main/default branch:

```
import pynautobot

nautobot = pynautobot.api(
    url = "http://0.0.0.0:8080",
    token = "nnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn"
)

# Change the dolt branch to "atl-leaf-augment"
nautobot.http_session.headers['dolt-branch'] = "atl-leaf-augment"

# Get a device that only exists on "atl-leaf-augment" branch
print(f"on 'atl-leaf-augment' branch, this device exists: {nautobot.dcim.devices.get(name='atl-leaf-09')}")

print()
print("Now change branch to 'main'")
nautobot.http_session.headers['dolt-branch'] = "main"

print()
print(f"On the 'main' branch, the same device does not exist: {nautobot.dcim.devices.get(name='atl-leaf-09')}")
```

Here is the output:
```
Documents/python_tests % python3 -i pynautobot_dolt_attempt.py
on 'atl-leaf-augment' branch, this device exists: atl-leaf-09

Now change branch to 'main'

On the 'main' branch, the same device does not exist: None
```

This is a fairly simple mod to make and it'd be quite easy to document.

Should pynautobot be made to accommodate the Version Control app more natively, or is this simple mod adequate?

**Comment by jvanderaa** [comment]
Is there an API endpoint that can be created within the Version Control app that can control this rather than updating pynautobot? Pynautobot already has the ability to interact with API endpoints of apps.

**Reply by tim-fiola** [reply]
I am not convinced that modifying pynautobot and/or Version Control makes anything simpler since you'd have to specify the desired `dolt-branch` somewhere (if querying something other than the `main` dolt-branch).  Pynautobot will query the `main` branch by default.

In a standard REST query against Nautobot Version Control you specify the `dolt-branch` in the header if you wish to query something other than the `main` branch.  It is a new parameter, but it is very simple to add this parameter to the query.  

I think it makes sense to carry this same pattern over to pynautobot (as in the initial code example) and just document it since it is very simple to implement that parameter in the session header:

```
nautobot = pynautobot.api(
    url = "http://0.0.0.0:8080",
    token = "nnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn"
)

# Change the dolt branch to "atl-leaf-augment"
nautobot.http_session.headers['dolt-branch'] = "atl-leaf-augment"
```

Leaving everything as-is, by default pynautobot will query the `main` branch. If you wish to query a non-main branch, simply specify that in the session header. This is an easy documentation fix.
